File: src/frontend/login/__/src/utils/csrf.js
import Cookies from 'js-cookie';
import { api } from '../main'; // Import the api instance from main.js

// Function to get CSRF token from cookies using js-cookie
function getCsrfToken() {
    return Cookies.get('csrftoken');
}

// Function to check if CSRF cookie is valid
async function checkCSRF() {
    const csrfToken = getCsrfToken();
  
    if (csrfToken) {
        try {
            const response = await api.get("/api/csrf/");
            if (response.data.status === "off") {
                console.log("CSRF token is invalid or expired");
                Cookies.remove('csrftoken', { path: '/' });
                return false;
            }
            api.defaults.headers.common['X-CSRFToken'] = response.headers["x-csrftoken"];
            return true;
        } catch (error) {
            console.error("Error checking CSRF token:", error);
            Cookies.remove('csrftoken', { path: '/' });
            return false;
        }
    }
    return false;
}

// Function to fetch and set CSRF token
export default async function fetchAndSetCsrfToken() {
    try {
        const status = await checkCSRF();
        if (status) {
            console.log('Valid CSRF tokens were found');
            return; // CSRF token is valid, no need to fetch a new one
        }
        console.log('No valid CSRF tokens were found');

        // Fetch new CSRF token from the server
        const response = await api.get("/api/csrf/");
        const csrfToken = response.headers["x-csrftoken"];

        // Set CSRF token in Axios defaults and cookies
        api.defaults.headers.common['X-CSRFToken'] = csrfToken;
    } catch (error) {
        console.error("Error fetching CSRF token:", error);
        throw error;
    }
}

File: src/backend/csrf/csrf_app/csrf.py
# csrf.py
from django.http import JsonResponse, HttpResponse
from django.middleware.csrf import get_token, CsrfViewMiddleware
from django.utils.decorators import method_decorator
from django.views.decorators.csrf import ensure_csrf_cookie, csrf_protect
from django.views import View
import requests

@csrf_protect
def check_csrf_token(request):
    csrf_token_from_cookie = request.COOKIES.get('csrftoken')
    csrf_token_from_header = request.META.get('HTTP_X_CSRFTOKEN')

    if not csrf_token_from_cookie:
        return JsonResponse({'status': 'off', 'message': 'No CSRF token found in cookies'})
    
    csrf_middleware = CsrfViewMiddleware(lambda req: None)

    try:
        csrf_middleware.process_view(request, None, (), {})
        return JsonResponse({'status': 'valid', 'csrftoken': csrf_token_from_cookie})
    except Exception:
        response = JsonResponse({'status': 'invalid', 'message': 'Invalid CSRF token'})
        response.delete_cookie('csrftoken')
        return response

@ensure_csrf_cookie
def get_csrf(request):
    response = JsonResponse({'detail': 'CSRF cookie set'})
    response['X-CSRFToken'] = get_token(request)
    return response

class ProxyView(View):
    @method_decorator(csrf_protect)
    def dispatch(self, *args, **kwargs):
        return super().dispatch(*args, **kwargs)

    def post(self, request, *args, **kwargs):
        return self.handle_request(request)

    def put(self, request, *args, **kwargs):
        return self.handle_request(request)

    def delete(self, request, *args, **kwargs):
        return self.handle_request(request)

    def patch(self, request, *args, **kwargs):
        return self.handle_request(request)

    def handle_request(self, request):
        csrf_middleware = CsrfViewMiddleware()
        if csrf_middleware.process_view(request, None, (), {}):
            return JsonResponse({'detail': 'Invalid CSRF token'}, status=403)

        service_url = self.get_service_url(request.path)
        if not service_url:
            return JsonResponse({'detail': 'Service URL not found'}, status=404)

        headers = self.get_headers(request)
        response = requests.request(
            method=request.method,
            url=service_url,
            headers=headers,
            data=request.body
        )

        return HttpResponse(response.content, status=response.status_code, content_type=response.headers.get('Content-Type'))

    def get_service_url(self, path):
        service_mapping = {
            '/api/pong/': 'http://pong:8000',
            # Add more mappings as needed
        }
        for prefix, target in service_mapping.items():
            if path.startswith(prefix):
                return target + path
        return None

    def get_headers(self, request):
        headers = {key: value for key, value in request.headers.items() if key.startswith('HTTP_')}
        headers['Content-Type'] = request.content_type
        return headers
